### DISCUSSION TITLE
Recursive retrieval of a dag

### DISCUSSION THREAD

**Discussion by 2color** [ipfs/helia#677] [Q&A] [answered]
Given a CID of a UnixFS director, e.g. [`bafybeiakerpabz5sbcl7e3sokxygunvtcj7lhyjl4me62uf2dxqzza5nha`](https://explore.ipld.io/#/explore/bafybeiakerpabz5sbcl7e3sokxygunvtcj7lhyjl4me62uf2dxqzza5nha), how would I recursively retrieve the whole DAG with Helia?

**Comment by achingbrain** [comment]
Could you specify a bit more detail about what you're trying to do?

Is there some sort of output format you want or do you just want to have the blocks locally?

**Reply by 2color** [reply]
This came up in the context of the Service Worker Gateway. 

If you want to make a site available offline, it's useful the have all the blocks in the blockstore.

**Reply by achingbrain** [reply, marked as answer]
You could pin it, then the whole DAG will be recursively fetched and the blocks will be in the blockstore.

**Reply by 2color** [reply]
Thanks.

Here's the code for anyone interested in what this looks like programatically 

https://github.com/ipfs/helia/blob/aecac3d92cbd22a7331afee8e6f87ef31a9f7d95/packages/utils/src/pins.ts#L71-L139

**Reply by achingbrain** [reply]
An nice feature might be using the pin name or pin metadata to let the user select previous versions of the site?  Kind of like the [Wayback Machine](https://web.archive.org/).

**Reply by 2color** [reply]
Yeah, that would be nice. It's a bit tricky adding a UI to the SW Gateway without interfering with the site backed by the CID you are viewing.
